target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c

```diff
@@ -1453,66 +1453,109 @@ static void rtl931x_set_distribution_algorithm(int group, int algoidx, u32 algom
 
 static void rtldsa_931x_led_init(struct rtl838x_switch_priv *priv)
 {
-	u64 pm_copper = 0, pm_fiber = 0;
+	u64 pm_copper = 0, pm_fiber = 0, pm_forced = 0;
 	struct device_node *node;
+	struct device *dev = priv->dev;
+	bool is_pm_auto_detected = false;
 
-	pr_debug("%s called\n", __func__);
 	node = of_find_compatible_node(NULL, NULL, "realtek,rtl9300-leds");
 	if (!node) {
-		pr_debug("%s No compatible LED node found\n", __func__);
+		dev_dbg(dev, "No compatible LED node found\n");
 		return;
 	}
 
+	for (int set = 0; set < 4; set++) {
+		char set_name[16] = {0};
+		u32 set_config[4];
+		int leds_in_this_set = 0;
+
+		/* Reset LED set configuration */
+		sw_w32(0, RTL931X_LED_SETX_0_CTRL(set));
+		sw_w32(0, RTL931X_LED_SETX_1_CTRL(set));
+
+		/*
+		 * Each LED set has (up to) 4 LEDs, and each LED is configured
+		 * with 16 bits. So each 32 bit register holds configuration for
+		 * 2 LEDs. Therefore, each set requires 2 registers for
+		 * configuring all 4 LEDs.
+		 */
+		snprintf(set_name, sizeof(set_name), "led_set%d", set);
+		leds_in_this_set = of_property_count_u32_elems(node, set_name);
+
+		if (leds_in_this_set <= 0 || leds_in_this_set > sizeof(set_config)) {
+			if (leds_in_this_set != -EINVAL) {
+				dev_err(dev, "%s invalid, skipping this set, leds_in_this_set=%d, should be (0, %d]\n",
+					set_name, leds_in_this_set, sizeof(set_config));
+			}
+
+			continue;
+		}
+
+		dev_info(dev, "%s has %d LEDs configured\n", set_name, leds_in_this_set);
+
+		if (of_property_read_u32_array(node, set_name, set_config, leds_in_this_set))
+			break;
+
+		/* Write configuration for selected LEDs */
+		for (int i = 0, led = leds_in_this_set - 1; led >= 0; led--, i++) {
+			sw_w32_mask(0xffff << RTL931X_LED_SET_LEDX_SHIFT(led),
+						(0xffff & set_config[i]) << RTL931X_LED_SET_LEDX_SHIFT(led),
+						RTL931X_LED_SETX_LEDY(set, led));
+		}
+	}
+
+	is_pm_auto_detected = of_property_read_u64(node, "realtek,forced-port-mask", &pm_forced);
```

Comment:
Do you need the return value at all here?
The pm_forced is ONLY updated if `realtek,forced-port-mask` is found (well.. if ret=0).
So as long as you defaulted pm_forced to 0 (i.e. no forced ports), then you could do without the somewhat confusing conditional logic further down...
It would instead just be
``` c
if ( !(priv->ports[i].phy || ( BIT_ULL(i) & pm_forced ) ) ) {
    continue;
}
```